### PR TITLE
feat: show quest action as complete after user has completed it

### PR DESCRIPTION
**Pull request by ribeirojose**
- add siwe auth example;
- start quest automatically after user starts their first action;
- refactor action tracking logic on the frontend to show actions as complete

### CODE REVIEW

File: apps/govquests-api/govquests/processes/lib/processes/start_quest_on_action_execution_started.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    private
+
+    def event_already_processed?(event)
+      @processed_event_ids[event.event_id]
+    end
+
+    def mark_event_as_processed(event)
+      @processed_event_ids[event.event_id] = true
+    end
```
Comment (ribeirojose) [inline apps/govquests-api/govquests/processes/lib/processes/start_quest_on_action_execution_started.rb:56]: not relevant anymore

File: apps/govquests-api/govquests/questing/lib/questing.rb [hunk 2]
```diff
@@ -3,13 +3,19 @@
...
 
 module Questing
   class Configuration
     def call(event_store, command_bus)
       command_handler = OnQuestCommands.new(event_store)
       command_bus.register(CreateQuest, command_handler)
       command_bus.register(AssociateActionWithQuest, command_handler)
+      command_bus.register(StartUserQuest, command_handler)
+      command_bus.register(CompleteUserQuest, command_handler)
+      command_bus.register(UpdateUserQuestProgress, command_handler)
```
Comment (ribeirojose) [inline apps/govquests-api/govquests/questing/lib/questing.rb:18]: remove

File: apps/govquests-api/govquests/questing/lib/questing.rb [hunk 3]
```diff
@@ -3,13 +3,19 @@
 require_relative "questing/events"
 require_relative "questing/on_quest_commands"
 require_relative "questing/quest"
+require_relative "questing/user_quest"
+
+# govquests/questing/lib/questing/configuration.rb
```
Comment (ribeirojose) [inline apps/govquests-api/govquests/questing/lib/questing.rb:8]: remove

File: apps/govquests-api/rails_app/app/graphql/mutations/generate_siwe_message.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    field :message, String, null: false
+    field :nonce, String, null: false
+
+    def resolve(address:, chain_id:)
+      nonce = ::Siwe::Util.generate_nonce
+      message = ::Siwe::Message.new(
+        "http://localhost:3000",
+        address,
+        "http://localhost:3000",
```
Comment (ribeirojose) [inline apps/govquests-api/rails_app/app/graphql/mutations/generate_siwe_message.rb:14]: hardcoded, change to env/secret

File: apps/govquests-api/rails_app/app/graphql/mutations/generate_siwe_message.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        "1",
+        {
+          statement: "SIWE Rails Example",
```
Comment (ribeirojose) [inline apps/govquests-api/rails_app/app/graphql/mutations/generate_siwe_message.rb:17]: `#must` update statement

File: apps/govquests-frontend/src/wagmi.ts [hunk 6]
```diff
@@ -1,15 +1,67 @@
...
-  transports: {
-    [optimism.id]: http(),
+export const config = createConfig(
+  getDefaultConfig({
+    appName: "GovQuests",
+    chains: [optimism],
+    transports: {
+      [optimism.id]: http(),
+    },
+    walletConnectProjectId: "ABXC",
```
Comment (ribeirojose) [inline apps/govquests-frontend/src/wagmi.ts:23]: bogus, fix this

File: apps/govquests-api/rails_app/app/read_models/questing/read_model_configuration.rb [hunk 7]
```diff
@@ -13,15 +13,15 @@ class QuestActionReadModel < ApplicationRecord
     belongs_to :action, class_name: "ActionTracking::ActionReadModel", foreign_key: "action_id", primary_key: "action_id"
   end
 
-  class UserQuest < ApplicationRecord
+  class UserQuestReadModel < ApplicationRecord
```
Comment (ribeirojose) [inline apps/govquests-api/rails_app/app/read_models/questing/read_model_configuration.rb:16]: this was generating a bunch of errors with code reloading 😓

File: apps/govquests-frontend/src/domains/authentication/hooks.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+  });
+};
+
+export const useSignOut = () => {
+  return useMutation({
+    mutationFn: signOut,
+  });
+};
+
+export const useCurrentUser = () => {
```
Comment (devjoaov) [inline apps/govquests-frontend/src/domains/authentication/hooks.ts:25]: 🔝

File: apps/govquests-api/rails_app/app/graphql/mutations/sign_in_with_ethereum.rb [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+    def resolve(signature:)
+      message = ::Siwe::Message.from_json_string(context[:session][:siwe_message])
+
+      if message.verify(signature, message.domain, message.issued_at, message.nonce)
+        context[:session][:siwe_message] = nil
+
+        user_id = SecureRandom.uuid
+
+        Rails.configuration.command_bus.call(Authentication::RegisterUser.new(
```
[the remark this answers is not in the document]
Comment (ribeirojose) [reply apps/govquests-api/rails_app/app/graphql/mutations/sign_in_with_ethereum.rb:16]: we should this should error out, good call 😬

**Review by luizmuraro — APPROVED** [review]

File: apps/govquests-frontend/src/domains/authentication/services/authService.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  return await request(API_URL, GENERATE_SIWE_MESSAGE, variables, {
+    credentials: "include",
+  });
+};
+
+export const signInWithEthereum = async (
+  variables: VariablesOf<typeof SIGN_IN_WITH_ETHEREUM>,
+) => {
+  return await request(API_URL, SIGN_IN_WITH_ETHEREUM, variables, {
+    credentials: "include",
```
Comment (luizmuraro) [inline apps/govquests-frontend/src/domains/authentication/services/authService.ts]: `#question` what exatcly is the objective of this credentials params?

**Review by luizmuraro — APPROVED** [review]

File: apps/govquests-frontend/src/domains/authentication/components/SignInButton.tsx [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+  if (!isConnected) {
+    return <ConnectKitButton />;
+  }
+
+  if (isSignedIn) {
+    return <Button onClick={() => signOut()}>Sign Out</Button>;
+  }
+
+  return <Button onClick={() => signIn()}>Sign-In with Ethereum</Button>;
+};
```
[the remark this answers is not in the document]
Comment (ribeirojose) [reply apps/govquests-frontend/src/domains/authentication/components/SignInButton.tsx]: when they're `connected`, they have explicitly allowed their wallet to connect to the site. After they're connected, they can opt to sign in. I think we covered this in our call, but lmk if you want a refresher!

File: apps/govquests-frontend/src/domains/authentication/services/authService.ts (hunk 10 above)
Comment (ribeirojose) [reply apps/govquests-frontend/src/domains/authentication/services/authService.ts]: I was trying to get the cookies working, but already figured out we don't need that 😬

File: apps/govquests-frontend/src/wagmi.ts (hunk 6 above)
Comment (ribeirojose) [reply apps/govquests-frontend/src/wagmi.ts:23]: OP-403

**Pull request merged**